sdk/lib/drivers/sound/mmebuddy/wave/streaming.c

```diff
@@ -206,6 +212,15 @@ CompleteIO(
             /* Partially completed */
             HdrExtension->BytesCompleted += dwNumberOfBytesTransferred;
             SND_TRACE(L"%d/%d bytes of wavehdr completed\n", HdrExtension->BytesCompleted, WaveHdr->dwBufferLength);
+
+            /* Do we loop a header? */
+            if (HdrExtension->BytesCommitted == WaveHdr->dwBufferLength &&
```

Comment:
To confirm, we want to check and reset `BytesCommitted`,
and we don't have to do anything with `BytesCompleted` or `dwNumberOfBytesTransferred`, for example?
For example, what does the above `SND_TRACE()` report?

Reply:
> we don't have to do anything with `BytesCompleted` or `dwNumberOfBytesTransferred`, for example?

No, we don't need to check, because `BytesCompleted` isn't used in a streaming loop (`DoWaveStreaming`). It only used for the two `SND_TRACE`s above and indicates completed bytes count instead. But it does not matter in case we don't use the completion. So we need to reset `BytesCommitted` instead, to continue the loop.
If you'll look to wdmaud.drv code, you'll see that `dwNumberOfBytesTransferred` is simply taken from `WaveHdr->dwBufferLength`, and it is same used only for two `SND_TRACE`s above.

> For example, what does the above `SND_TRACE()` report?

As I tested, it shows `BytesCompleted/WaveHdr->dwBufferLength`. But in case completion is not used, it indicates the total committed buffer size (e. g., `WaveHdr->dwBufferLength` multiplied to played amount of loops). So it has no any other purpose in this case, excepting displaying the total committed buffer size. 2nd counter always displays the length of one wave header.

Reply:
This is what I assumed, thanks for the confirmation.
Then, the 3 'Partially completed' lines should be moved to an `else` block, to clarify/optimize.

Reply:
Done.

Reply:
No, because we don't use this in case of looping, when we don't use the header completion. And even in case of completion, we don't need to reset it, because it's actually used as a total completed bytes count in pair with `dwNumberOfBytesTransferred`, which is used as a condition for a completion loop, and we are storing it for other purposes here, so it cannot be `0` due to this at least:
https://git.reactos.org/?p=reactos.git;a=blob;f=sdk/lib/drivers/sound/mmebuddy/wave/header.c;hb=5949c20d97adf135cefaadba46b887391f31030e#l359

Reply:
Thanks. I'll trust you, as I don't want to dig this code further at the moment.